### ISSUE TITLE
How does one remove all logical white spaces

### ISSUE THREAD

**Issue by a1730** [casid/jte#80]
Hello,
Is it possible to remove **all** logical white spaces in the final Jte output? We are using Jte for XML templating and it insists on leaving logical white spaces in our final XML documents thereby making the document invalid for some XML processors.
Suppose I have the following template:
```
@import com.example.JteXMLIssue
@import com.example.HttpStatus
@import java.time.Instant
@import java.time.temporal.ChronoUnit
@param JteXMLIssue res
!{
    var ts = Instant.now().truncatedTo(ChronoUnit.MILLIS).toString();
    var message = res.message().isBlank()? HttpStatus.stringOf(200).toUpperCase(): res.message();
}
<!DOCTYPE cXML SYSTEM "http://example.com/example.dtd">
<Issue xml:lang="en-US" timestamp="${ts}" someID="${res.pid()}@internal.example.com">
  <Response><Status code="${res.code()}" text="${HttpStatus.stringOf(res.code())}">${message}</Status></Response>
</Issue>
```
Where JteXMLIssue  is defined as `public record JteXMLIssue(int code, String pid, String message) {}`
The template works.
However, Jte leaves leading lines in the response, and insists on reformatting the document like this:
```
=========blank line==========
<!DOCTYPE cXML SYSTEM "http://example.com/example.dtd">
<Issue xml:lang="en-US" timestamp="2021-05-06T08:46:07.797Z" someID="53puf4mq@internal.example.com">
  <Response>
    <Status code="200" text="OK">OK</Status>
  </Response>
</Issue>
``` 
Reformatting is not a problem but the leading white spaces makes the document invalid for some XML readers.
We tried putting the DOCTYPE line before the code block but Jte compiler complained.
What can one do to remove all logical white spaces, or how do I get Jte to render something like the following with no leading white spaces or new lines:
```<!DOCTYPE cXML SYSTEM "http://example.com/example.dtd">
<Issue xml:lang="en-US" timestamp="2021-05-06T08:46:07.797Z" someID="53puf4mq@internal.example.com">
  <Response><Status code="200" text="OK">OK</Status></Response>
</Issue>
```
The following output will even be better:
`<!DOCTYPE cXML SYSTEM "http://example.com/example.dtd">
<Issue xml:lang="en-US" timestamp="2021-05-06T08:46:07.797Z" someID="53puf4mq@internal.example.com">
  <Response><Status code="200" text="OK">OK</Status></Response>
</Issue>
`
Thank you for an excellent project.

**Comment by casid** [comment]
@a1730 thanks for the heads up!

I think the whitespace at the beginning should be solved by enabling this setting:
`gg.jte.TemplateEngine#setTrimControlStructures`

Here's a unit test describing what the setting does :-)
https://github.com/casid/jte/blob/master/jte/src/test/java/gg/jte/TemplateEngine_TrimControlStructuresTest.java

There's also this open issue #10. I think this is going in the direction you would like as ideal output :-). I still haven't found the time to implement it though, since it is especially tricky for HTML.

Let me know if `setTrimControlStructures` works for the parsing issues. If so, I'd tend to close this issue and add a reference to it at #10.

**Comment by casid** [comment]
@a1730 you're right, I can reproduce your finding with `setTrimControlStructures=true`.

As a quick workaround you can rearrange your variable declarations like this:

```
@import com.example.JteXMLIssue
@import com.example.HttpStatus
@import java.time.Instant
@import java.time.temporal.ChronoUnit
@param JteXMLIssue res
<!DOCTYPE cXML SYSTEM "http://example.com/example.dtd">
!{
    var ts = Instant.now().truncatedTo(ChronoUnit.MILLIS).toString();
    var message = res.message().isBlank()? HttpStatus.stringOf(200).toUpperCase(): res.message();
}
<Issue xml:lang="en-US" timestamp="${ts}" someID="${res.pid()}@internal.example.com">
  <Response><Status code="${res.code()}" text="${HttpStatus.stringOf(res.code())}">${message}</Status></Response>
</Issue>
```

I'm currently investigating why the variable section doesn't get trimmed correctly when at the beginning of the file.

Although, if you have the same `<!DOCTYPE` structure in many files, maybe a layout could be useful as well. Something like this:

layout/document.jte
```
@param gg.jte.Content content
<!DOCTYPE cXML SYSTEM "http://example.com/example.dtd">
${content}
```

response.jte
```
@import com.example.JteXMLIssue
@import com.example.HttpStatus
@import java.time.Instant
@import java.time.temporal.ChronoUnit
@param JteXMLIssue res
@layout.document(content = @`
!{
    var ts = Instant.now().truncatedTo(ChronoUnit.MILLIS).toString();
    var message = res.message().isBlank()? HttpStatus.stringOf(200).toUpperCase(): res.message();
}
<Issue xml:lang="en-US" timestamp="${ts}" someID="${res.pid()}@internal.example.com">
  <Response><Status code="${res.code()}" text="${HttpStatus.stringOf(res.code())}">${message}</Status></Response>
</Issue>
`)
```

You can add more to the layout and also parameterize it, for instance if the `<Issue` part is the same for all responses as well.

**Comment by casid** [comment]
@a1730 I submitted a fix for your issue, the `!{}` block is now correctly trimmed as first control structure.

**Comment by a1730** [comment]
Hello @casid ,
Awesome! We patiently await the GA release.
Furthermore, we appreciate the information and advice regarding template structural arrangement and use of  layouts.  I should probably note that  we naturally went for layouts  in some of the code base that are being migrated from JSPs.
However, we trimmed down the issue report to create [a minimal, complete and verifiable example](https://stackoverflow.com/help/minimal-reproducible-example) for you.

We really appreciate your responsiveness and fast turn-around on this issue.
Thank you.

**Comment by casid** [comment]
I'm planning to release a new jte version this weekend, including this fix.

Also, you're migrating JSPs to jte? We're currently evaluating this as well at work. And there's a converter that should help a lot with the boilerplate work of doing this: https://github.com/casid/jte/tree/master/jte-jsp-converter

Unfortunately I haven't written any documentation for it yet, but there are some tests to demonstrate the usage. You can easily extend the converter with your custom tags used in your project, by implementing a main method with the converter. With the IntelliJ plugin you can then navigate to a jsp file and click `convert JSP to jte` 👍 

The converter fails if you attempt to convert a JSP that uses other JSPs and will tell you to convert those first. At work, I've written a briding tag, that can be used to call jte tags from JSP templates, so you can do a incremental replacement without the need for long lived branches and duplicate code.

A bit off-topic, but maybe it helps your migration!

**Comment by casid** [comment]
@a1730 I just released `1.10.0` to maven central.

I'm closing this issue now, feel free to reopen if there are still problems afterwards!
